src/query/expression/src/filter/select_expr_permutation.rs

```diff
@@ -0,0 +1,112 @@
+// Copyright 2021 Datafuse Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use rand::Rng;
+
+#[derive(Clone, Debug)]
+pub struct FilterPermutation {
+    // Whether we can reorder filter expressions.
+    can_reorder: bool,
+    // Whether there is a swap to be observed.
+    observe: bool,
+    // If there is a better swap, we keep it at least once.
+    keep_better: bool,
+    // The last swap index is swap_idx and swap_idx + 1.
+    swap_idx: usize,
+    // The sum of runtime of this permutation.
+    runtime_sum: u64,
+    // The count of runtime of this permutation.
+    runtime_count: u64,
+    // The possibility of swap for index [0, permutation.len() - 1]
+    swap_possibility: Vec<u32>,
+    // The permutation of filter expressions.
+    permutation: Vec<usize>,
+    // The border of random number.
+    random_number_border: u64,
+}
+
+impl FilterPermutation {
+    pub fn new(num_filter_exprs: usize, can_reorder: bool) -> Self {
+        Self {
+            can_reorder,
+            observe: false,
+            keep_better: false,
+            swap_idx: 0,
+            runtime_sum: 0,
+            runtime_count: 0,
+            swap_possibility: vec![100; num_filter_exprs - 1],
+            permutation: (0..num_filter_exprs).collect(),
+            random_number_border: 100 * (num_filter_exprs - 1) as u64,
+        }
+    }
+
+    pub fn get(&self, idx: usize) -> usize {
+        if self.can_reorder {
+            self.permutation[idx]
+        } else {
+            idx
+        }
+    }
+
+    pub fn add_statistics(&mut self, runtime: u64) {
+        if !self.can_reorder {
+            return;
+        }
+
+        // Whether there is a swap to be observed.
+        if self.observe {
```

Comment:
Do we need to add a `iteration_count`?

Reply:
No, I have made some improvements and tested them in our model, the result show that our algorithm is better.